roachprod: add A records during cluster create

dt opened the pull request:
Release note: none.
Epic: none.

dt committed:
roachprod: add A records during cluster create

Release note: none.
Epic: none.

golgeek reviewed (commented):
We should probably also delete the DNS records in `roachprod.destroyCluster()` to avoid having dangling DNS records for destroyed clusters.

dt commented:
> delete the DNS records

At first I thought we already did, since `destroy` prints this:

```
2025/07/11 15:25:18 roachprod.go:1577: Destroying cluster david-test with 1 nodes
2025/07/11 15:25:18 client.go:310: Prometheus credentials obtained from File
Destroying Prometheus configs
Destroying DNS entries
Destroying VMs
...
```

But on closer inspection I realized is only destroying the (non-existent) SRV records in the managed domain, not any A records ore records in the public zone, since the helper functions we have for finding and deleting records are currently specialized to that zone/those records.

That said, I believe the next `sync` or `list` that anyone runs, including the frequent gc cron job, will clean up any orphaned records since it rebuilds the whole zone file, so they records for a destroyed cluster, even if we do nothing here, should be cleaned up within an hour or so, right? Given that do you still think it is worth adding code to destroy them eagerly during destroy?

herkolategan commented:
> > That said, I believe the next `sync` or `list` that anyone runs, including the frequent gc cron job, will clean up any orphaned records since it rebuilds the whole zone file, so they records for a destroyed cluster, even if we do nothing here, should be cleaned up within an hour or so, right? Given that do you still think it is worth adding code to destroy them eagerly during destroy?
> 
> From what I can see, the `dnsProvider.CreateRecords()` func manages creating both A and SRV records, but the "create or update" logic only works for SRV (it only lists SRV to check if the record already exists or not). I think that would break the re-creation of a cluster right it was destroyed if no sync/gc ran in between.

Yep, these were originally only written to support SRV records.

dt commented:
I probably won't be able to revisit this this week (my local `roachprod` binary is maintaining my A records now so I've moved on to my planned benchmarking work) so if you'll get to it before me, please feel free to take over this PR; I'll be sure to check in before force-pushing if/when this manages to get back on my queue.

golgeek committed:
roachprod: add A records during cluster create

Patch the lookup logic to handle A records in addition to SRV records
and make the update/create record logic work with public records.

Epic: none
Release note: None

dt commented:
LGTM!

did you take it for a manual spin or should I give it one more go before I bors?

dt commented:
I built a binary with the current tip and created a couple clusters all of which had dns despite my expired AWS SSO, so looks great!

Thanks for getting this over the line!

bors r+

golgeek commented:
> ah, yeah, I guess we still need a green check.

Not sure how kosher it is that I approve, but we now have a green checkmark!

> did you take it for a manual spin or should I give it one more go before I bors?

> I built a binary with the current tip and created a couple clusters all of which had dns despite my expired AWS SSO, so looks great!

👍

herkolategan commented on pkg/roachprod/cloud/gc.go:
```diff
@@ -570,7 +570,7 @@ func GCDNS(l *logger.Logger, cloud *Cloud, dryrun bool) error {
 
 		if err := destroyResource(dryrun, func() error {
-			return p.DeleteRecordsByName(ctx, recordNames...)
+			return p.DeleteSRVRecordsByName(ctx, recordNames...)
```
During GC we look for DNS records that now longer pertain to an existing cluster. This generally only happens if something goes wrong during a cluster delete and the DNS records remain for some reason.

It's an edge case. But we should probably handle the A records here as well then.

herkolategan replied:
Ah, if the rewrite happens during GC that should be fine! LGTM then.

herkolategan reviewed (approved):
Other than considering GC for dangling A records, this PR LGTM.

golgeek commented:
Will use this feature in another PR, so borsing it; hope you don't mind @dt!

bors r+
